index.js

```diff
@@ -1,92 +1,117 @@
   return fdf;
-}
+};
 
 exports.generatePdf = function(data, templatePath, extendArgs, callback) {
-  var tempNameResult = temp.path({suffix: '.pdf'}),
+  let tempNameResult = temp.path({suffix: '.pdf'}),
       pdfPath        = isAbsolute(templatePath) ? templatePath : path.join(__dirname, templatePath);
 
+  let normalized = normalizeArgs(extendArgs, callback);
+
+  extendArgs = normalized.args;
+  callback   = normalized.callback;
+
+  let processArgs = [pdfPath, 'fill_form', '-', 'output', tempNameResult].concat(extendArgs);
+
+  let child = spawn('pdftk', processArgs);
+
+  writeFdfToPdftk(child, data);
+  handlePdftkError(child, callback);
+  handlePdftkExit(child, tempNameResult, callback);
+};
+
+function normalizeArgs(extendArgs, callback) {
   // Check if extendArgs is our callback, adds backwards compat
   if (typeof extendArgs === 'function') {
    callback = extendArgs;
    extendArgs = [];
   }
-  else if (extendArgs instanceof Array) {
-    args = extendArgs;
-  }
-  else {
+  else if (!(extendArgs instanceof Array)) {
     extendArgs = [];
   }
+  return { args: extendArgs, callback: callback };
+}
 
-  var processArgs = [pdfPath, "fill_form", "-", "output", tempNameResult].concat(extendArgs);
-
-  child = spawn("pdftk", processArgs);
-
-  child.on('exit', function(code) {
-
-    // If a exit code besides 0 was thrown then send an error
-    if ( code ) {
-      callback(new Error('Non 0 exit code from pdftk spawn: ' + code));
-    }
-
-    // Async read tempfile, we should think about making this a stream instead of a read into memory.
-    fs.readFile(tempNameResult, function(err, filledPdf) {
-
-      if ( err ) {
-        return callback(err);
-      }
-
-      // Delete files, doing nested callbacks for now, but lets look into maybe using async for this
-      fs.unlink(tempNameResult, function(err) {
-        if ( err ) {
-          return callback(err);
-        }
-
-        // Everything Looks good, send back pdfFile.
-        callback(null, filledPdf);
-      });
-    });
-  });
-
-  // Write FDF file to pdftk spawned process
+function writeFdfToPdftk(child, data) {
   child.stdin.write(exports.generateFdf(data));
   child.stdin.end();
+}
 
+function handlePdftkError(child, callback) {
   child.on('error', function (err) {
     callback(err);
   });
 
   child.stderr.on('data', function (data) {
     console.error('stderr: ' + data);
   });
+}
+
+function handlePdftkExit(child, tempNameResult, callback) {
+  child.on('exit', createHandlePdftkExit(tempNameResult, callback));
+}
+
+function createHandlePdftkExit(tempNameResult, callback) {
+  return function(code) {
+    if ( code ) {
+      return callback(new Error('Non 0 exit code from pdftk spawn: ' + code));
+    }
+
+    async.waterfall([
+      (cb) => {
+        fs.readFile(tempNameResult, (err, filledPdf) => cb(err, filledPdf));
+      },
+      (cb, filledPdf) => {
+        fs.unlink(tempNameResult, function(err) {
+          callback(err, filledPdf);
```

Comment:
Fixed! Thanks for the catch. That's what I get for trying to do some coding on my day off with 3 kids running around. lol